support general typed string literals

maxcountryman opened the pull request:
This is a direct port of MaterializeInc/materialize#3146. It provides
support for e.g. Postgres syntax where any string literal may be
preceded by a type name.

Fixes #168.

nickolay commented:
I wanted to note that even though that Materialize as a whole is licensed under a non-opensource license, they kept the sql-parser bits we borrow these changes from ([value.rs](https://github.com/benesch/materialize/blob/77811560a23cda4253844ed157ad119240d4cf9a/src/sql-parser/src/ast/value.rs) and [parser.rs](https://github.com/benesch/materialize/blob/77811560a23cda4253844ed157ad119240d4cf9a/src/sql-parser/src/parser.rs)) available under the Apache license, so there shouldn't be a problem with this.

maxcountryman commented:
Thank you for the feedback! I'll try to have another look at this later on this week.

maxcountryman commented on src/parser.rs:
```diff
@@ -1925,30 +1971,20 @@ impl Parser {
-            }
+            // Check if the recently consumed '(' started a derived table, in which case we've
+            // parsed the subquery, followed by the closing ')', and the alias of the derived
+            // table. In the example above this is case (3), or another nested join (2).
```
Thanks. I think I probably had copied this from the original MaterializeInc PR without thoroughly reading it--sorry about that.

maxcountryman commented on src/parser.rs:
```diff
@@ -34,6 +34,15 @@ macro_rules! parser_err {
 }
 
+// Returns a successful result if the optional expression is some
+macro_rules! return_ok_if_some {
```
Oh sure, happy to do it that way if you prefer!

maxcountryman commented on src/parser.rs:
```diff
@@ -176,6 +185,36 @@ impl Parser {
+        // expression that should parse as the column name "date".
+        return_ok_if_some!(self.maybe_parse(|parser| match parser.parse_data_type()? {
+            DataType::Interval => parser.parse_literal_interval(),
+            // Single-quoted strings are parsed as custom data types, however this not desirable
+            // when we are handling input like `"NOT 'a' NOT LIKE 'b'"` because this will produce a
+            // TypedString instead of a SingleQuotedString. Further, this leads to issues where the
+            // same input will yield a BinaryOperator instead of the correct UnaryOperator. Here we
+            // handle that specific case by returning an error.
+            DataType::Custom(..) => Err(ParserError::ParserError("".to_string())),
```
Okay. That does make sense. That said, I really don't love my implementation here.

maxcountryman replied:
Oh it seems like Materialize has [fully removed it from `DataType`](https://github.com/MaterializeInc/materialize/blob/eacdc6564b69afa0e37135d1bfb77b58dc6f4f96/src/sql-parser/src/ast/data_type.rs#L27-L77), so I'm going to assume that's what you mean here--I can update accordingly. 🙂

nickolay replied:
No, we probably should support parsing custom datatypes in other contexts, where this does not lead to ambiguity. But we can disallow custom types in `type 'str'`, like your patch now does, until someone needs it.

nickolay replied:
Also: I meant to suggest `DataType::Custom(..) => parser_err!("dummy"), but I don't see if I posted it elsewhere.

maxcountryman replied:
Got it, thanks for clarifying (I realized after that that removing Custom would break other things). I'll update the error accordingly.

maxcountryman commented on src/ast/mod.rs:
```diff
@@ -205,6 +205,8 @@ pub enum Expr {
     Nested(Box<Expr>),
     /// A literal value, such as string, number, date or NULL
     Value(Value),
+    /// A structure containing a `String` as well as an associated data type.
+    TypedString { data_type: DataType, value: String },
```
Thanks. This is much better!

maxcountryman commented on src/parser.rs:
```diff
@@ -176,6 +185,36 @@ impl Parser {
+            // Single-quoted strings are parsed as custom data types, however this not desirable
+            // when we are handling input like `"NOT 'a' NOT LIKE 'b'"` because this will produce a
+            // TypedString instead of a SingleQuotedString. Further, this leads to issues where the
+            // same input will yield a BinaryOperator instead of the correct UnaryOperator. Here we
```
I can remove this, it's just explaining why our test would fail if we didn't explicitly handle `DataType::Custom` by returning an error (e.g. it will parse as a BinaryOperator instead of an UnaryOperator).

maxcountryman replied:
Would it be better to change this entire comment to say something to the effect of we don't accept the type 'string' syntax for the custom data types? (Borrowing from your comment below.)

nickolay replied:
> It will parse as a BinaryOperator instead of an UnaryOperator
I don't understand _what_ will parse as BinaryOperator.

> Would it be better to change this entire comment to say something to the effect of we don't accept the type 'string' syntax for the custom data types? (Borrowing from your comment below.)

I think both the description of what returning an `Err()` here does and the non-obvious example of `NOT 'x'  LIKE 'y'` should stay in the comment. I'm just trying to make sure I understand the examples and that they are clearly described.

maxcountryman replied:
> I don't understand what will parse as BinaryOperator.

Here I mean the direct example given in the comment of `"NOT 'a' NOT LIKE 'b'"`.

> I think both the description of what returning an Err() here does and the non-obvious example of NOT 'x' LIKE 'y' should stay in the comment. I'm just trying to make sure I understand the examples and that they are clearly described.

Sure, makes sense. I may be approaching the limits of my (admittedly shallow) understanding of the library here so I may need to defer to you for how you'd like to document that.

nickolay replied:
> Here I mean the direct example given in the comment of "NOT 'a' NOT LIKE 'b'".

Ah, sorry for being dense.

maxcountryman commented on src/parser.rs:
```diff
@@ -1925,30 +1974,26 @@ impl Parser {
-                    Ok(TableFactor::NestedJoin(Box::new(table_and_joins)))
-                }
-            }
+            // Check if the recently consumed '(' started a derived table, in which case we've
+            // parsed the subquery, followed by the closing ')', and the alias of the derived
+            // table. In the example above this is case (3).
+            //
+            // A parsing error from `parse_derived_table_factor` indicates that the '(' we've
+            // recently consumed does not start a derived table (cases 1, 2, or 4). Ignore the
+            // error and back up to where we after the opening '('.
```
Sure, that would be helpful if you don't mind. 🙂

maxcountryman commented:
I rebased this off of master (and squashed to avoid having to resolve the merge conflicts in each commit) so this should now incorporate the mainline changes.

nickolay commented:
> I rebased this off of master (and squashed to avoid having to resolve the merge conflicts in each commit) so this should now incorporate the mainline changes.

Excellent.

I pushed an update to the comments, let me know if they make sense, and I think this is ready to merge?

maxcountryman commented on src/parser.rs:
```diff
@@ -172,6 +181,40 @@ impl Parser {
+        return_ok_if_some!(self.maybe_parse(|parser| {
+            match parser.parse_data_type()? {
+                DataType::Interval => parser.parse_literal_interval(),
+                // PosgreSQL allows almost any identifier to be used as custom data type name,
+                // and we support that in `parse_data_type()`. But unlike Postgres we don't
+                // have a list of globally reserved keywords (since they vary across dialects),
+                // so given `NOT 'a' LIKE 'b'`, we'd accept `NOT` as a possible custom data type
+                // name, resulting in `NOT 'a'` being recognized as a `TypedString` instead of
+                // an unary negation `NOT ('a' LIKE 'b')`. To solve this, we don't accept the
+                // `type 'string'` syntax for the custom data types at all.
```
This reads so much better. Thank you!

maxcountryman commented:
> I pushed an update to the comments, let me know if they make sense, and I think this is ready to merge?

Amazing. Thank you! It looks good to me. Really appreciate all the detailed feedback and of course patience. 🙂

nickolay commented:
Thank you as well!
